lib/pharos/addon.rb

```diff
@@ -26,62 +40,98 @@ def self.messages
+
+      # @return [String]
+      def addon_location
+        @addon_location || __dir__
       end
-    end
 
-    def self.version(version = nil)
-      if version
-        @version = version
-      else
-        @version
+      def name(name = nil)
+        if name
+          @name = name
+        else
+          @name
+        end
       end
-    end
 
-    def self.license(license = nil)
-      if license
-        @license = license
-      else
-        @license
+      def version(version = nil)
+        if version
+          @version = version
+        else
+          @version
+        end
       end
-    end
 
-    def self.to_h
-      { name: name, version: version, license: license }
-    end
+      def license(license = nil)
+        if license
+          @license = license
+        else
+          @license
+        end
+      end
 
-    def self.schema(&block)
-      @schema = Dry::Validation.Form(Schema, &block)
-    end
+      def to_h
+        { name: name, version: version, license: license }
+      end
 
-    def self.struct(&block)
-      @struct ||= Class.new(Pharos::Addons::Struct, &block)
-    end
+      def config_schema(&block)
+        @schema = Dry::Validation.Form(Schema, &block)
+      end
 
-    def self.validation
-      Dry::Validation.Form(Schema) { yield }
-    end
+      def config(&block)
+        @config ||= Class.new(Pharos::Addons::Struct, &block)
+      end
 
-    def self.validate(config)
-      if @schema
-        @schema.call(config)
-      else
-        validation {}.call(config)
+      def config?
+        !@config.nil?
+      end
+
+      def custom_type(&block)
+        Class.new(Pharos::Addons::Struct, &block)
       end
-    end
 
-    def self.descendants
-      ObjectSpace.each_object(Class).select { |klass| klass < self }
+      # @return [Hash]
+      def hooks
+        @hooks ||= {}
+      end
+
+      def install(&block)
+        hooks[:install] = block
+      end
+
+      def uninstall(&block)
+        hooks[:uninstall] = block
+      end
+
+      def validation
+        Dry::Validation.Form(Schema) { yield }
+      end
+
+      # @param config [Hash]
+      def validate(config)
+        if @schema
+          @schema.call(config)
+        else
+          validation {}.call(config)
+        end
+      end
+
+      def descendants
+        ObjectSpace.each_object(Class).select { |klass| klass < self }
+      end
     end
 
-    attr_reader :config, :cpu_arch, :cluster_config
+    attr_reader :config, :cpu_arch, :cluster_config, :master
 
+    # @param config [Hash]
+    # @param enabled [Boolean]
+    # @param master [Pharos::Configuration::Host,NilClass]
+    # @param cpu_arch [String, NilClass]
+    # @param cluster_config [Pharos::Config, NilClass]
     def initialize(config = nil, enabled: true, master:, cpu_arch:, cluster_config:)
-      @config = self.class.struct.new(config)
+      @config = self.class.config? ? self.class.config.new(config) : RecursiveOpenStruct.new(config)
```

Comment:
If we're directly using some dependency of a dependency like `RecursiveOpenStruct`, maybe it should be in our gemspec.